api/clients/codecs/encoded_payload.go

```diff
@@ -0,0 +1,158 @@
+
+	payloadBytes := payload.GetBytes()
+
+	// encode payload length as uint32
+	binary.BigEndian.PutUint32(
+		encodedPayloadHeader[2:6],
+		uint32(len(payloadBytes))) // uint32 should be more than enough to store the length (approx 4gb)
+
+	// encode payload modulo bn254, and align to 32 bytes
+	encodedData := codec.PadPayload(payloadBytes)
+	encodedPayloadBytes := append(encodedPayloadHeader, encodedData...)
+
+	return &encodedPayload{encodedPayloadBytes}, nil
+}
+
+// decode applies the inverse of PayloadEncodingVersion0 to an encodedPayload, and returns the decoded Payload
+func (ep *encodedPayload) decode() (*Payload, error) {
+	claimedLength := binary.BigEndian.Uint32(ep.bytes[2:6])
+
+	// decode raw data modulo bn254
+	unpaddedData, err := codec.RemoveInternalPadding(ep.bytes[32:])
+	if err != nil {
+		return nil, fmt.Errorf("remove internal padding: %w", err)
+	}
+
+	unpaddedDataLength := uint32(len(unpaddedData))
+
+	// data length is checked when constructing an encoded payload. If this error is encountered, that means there
+	// must be a flaw in the logic at construction time (or someone was bad and didn't use the proper construction methods)
+	if unpaddedDataLength < claimedLength {
+		return nil, fmt.Errorf(
+			"length of unpadded data %d is less than length claimed in encoded payload header %d. this should never happen",
+			unpaddedDataLength, claimedLength)
+	}
+
+	// unpadded data length can be slightly bigger than the claimed length, since RemoveInternalPadding doesn't
+	// do anything to remove trailing zeros that may have been added when the data was initially padded.
+	// however, this extra padding shouldn't exceed 31 bytes, because that's the most that would be added
+	// when padding the data length to 32 bytes. If this error occurs, that means there must be a flaw in the logic at
+	// construction time (or someone was bad and didn't use the proper construction methods)
+	if unpaddedDataLength > claimedLength+31 {
+		return nil, fmt.Errorf(
+			"length of unpadded data %d is more than 31 bytes longer than claimed length %d. this should never happen",
+			unpaddedDataLength, claimedLength)
+	}
+
+	return NewPayload(unpaddedData[0:claimedLength]), nil
+}
+
+// toFieldElements converts the encoded payload to an array of field elements
+func (ep *encodedPayload) toFieldElements() ([]fr.Element, error) {
+	fieldElements, err := rs.ToFrArray(ep.bytes)
+	if err != nil {
+		return nil, fmt.Errorf("deserialize field elements: %w", err)
+	}
+
+	return fieldElements, nil
+}
+
+// encodedPayloadFromElements accepts an array of field elements, and converts them into an encoded payload
+//
+// blobLength is the length of the blob IN SYMBOLS, as claimed by the blob commitment. This is needed to make sure
+// that the claimed length in the encoded payload header is valid relative to the total blob length
+func encodedPayloadFromElements(fieldElements []fr.Element, blobLength uint32) (*encodedPayload, error) {
```

Comment:
this comment makes me wonder whether we shouldn't instead define this method on the blob itself? Why do we need this method here? Seems like its at the wrong place given that it requires a blobLength?

Reply:
In [617faa7b63](https://github.com/Layr-Labs/eigenda/pull/1120/commits/617faa7b638785ebbcfc71d02ebdc5e6d0795cf6#diff-1556913a0af60ea5bf48b4b5a78d03e1be5575d485dcff7658fcadfde31a211f), I changed the function signature to the following, to remove blob related logic from `encodedPayload`:

```
encodedPayloadFromElements(fieldElements []fr.Element, maxPayloadLength uint32)
```